go/fn/runnerProcessor.go

```diff
@@ -30,45 +31,88 @@ type runnerProcessor struct {
+		return false
+	}
+	return o.GetKind() == "ConfigMap" && o.GetName() == "function-input" && len(data) == 0
+}
+
+// Process assigns the ResourceList.FunctionConfig to Runner's attributes, and calls the Runner.Run methods (main method)
+// to run functions. The r.fnRunner accepts three kinds of functionConfig value:
+// 1. no function config, it only runs fnRunner.Run
+// 2. ConfigMap type, it requires the Runner instance to have one contributes of type map[string]string to receive the ConfigMap `.data` value.
+// 3. Runner type, it uses the Runner struct name as the FunctionConfig Kind. e.g. if the Runner is `SetNamespace`,
+//   the FunctionConfig should be `{"Kind": "SetNamespace", "apiVersion": "fn.kpt.dev/v1alpha1"}
 func (r runnerProcessor) Process(rl *ResourceList) (bool, error) {
+	// Validate and Parse the input FunctionConfig to r.fnRunner
+	if rl.FunctionConfig.IsEmpty() || EmptyFunctionConfig(rl.FunctionConfig){
+		// functions may not need functionConfig.
+		rl.Results.Infof("`FunctionConfig` is not given")
+	} else {
+		err := r.config(rl.FunctionConfig)
+		if err != nil {
+			rl.Results.ErrorE(err)
+			return false, nil
+		}
+	}
+	// Run the main function.
 	fnCtx := &Context{Context: r.ctx}
 	results := new(Results)
-	if ok := r.config(rl.FunctionConfig, results); !ok {
-		rl.Results = append(rl.Results, *results...)
-		return false, nil
-	}
-	pass := r.fnRunner.Run(fnCtx, rl.FunctionConfig, rl.Items, results)
+	shouldPass := r.fnRunner.Run(fnCtx, rl.FunctionConfig, rl.Items, results)
+	// If running in a pipeline, the ResourceList may already have results from previous function runs.
+	// Thus, we only append new results to the end.
 	rl.Results = append(rl.Results, *results...)
-	return pass, nil
+	return shouldPass, nil
 }
 
-func (r *runnerProcessor) config(o *KubeObject, results *Results) bool {
-	fnName := reflect.ValueOf(r.fnRunner).Elem().Type().Name()
-	switch true {
-	case o.IsEmpty():
-		results.Infof("`FunctionConfig` is not given")
-	case o.IsGroupKind(schema.GroupKind{Kind: "ConfigMap"}):
-		data, _, err := o.NestedStringMap("data")
-		if err != nil {
-			results.ErrorE(err)
-			return false
-		}
-		fnRunnerElem := reflect.ValueOf(r.fnRunner).Elem()
-		for i := 0; i < fnRunnerElem.NumField(); i++ {
-			if fnRunnerElem.Field(i).Kind() == reflect.Map {
-				fnRunnerElem.Field(i).Set(reflect.ValueOf(data))
-				break
-			}
-		}
-	case o.IsGroupVersionKind(schema.GroupVersionKind{Group: "fn.kpt.dev", Version: "v1alpha1", Kind: fnName}):
-		err := o.As(r.fnRunner)
-		if err != nil {
-			results.ErrorE(err)
-			return false
+func (r *runnerProcessor) config(o *KubeObject) error {
+	if r.fnRunner == nil {
+		return fmt.Errorf("the object which implements `ResourceListProcessor` interface requires a `Runner` or `fnRunner` attribute," +
+			" got nil")
+	}
+	switch o.GroupKind() {
+	case schema.GroupKind{Kind: "ConfigMap"}:
+		data, found, err := o.NestedStringMap("data")
+		if !found || err != nil {
```

Comment:
Q: Could you could handle empty data here, instead of the precheck?

Reply:
Are you asking in general whether we can skip the "found" or just for this case?
For this case, yes, they can be simplified as 
```
if data == nil {
      return err
}
``` 
In general, I think it will miss some info users would really want to know,  because only "found" can differentiate the existence of an empty field. e.g.  `o.GetNestedStringMap("key1")`  
```
# found == false, data == nil
key1: {}
key2: 100 
``` 
and 
```
# found == true, data == nil
key2: 100 
```
For cases like an empty field value is accepted, but the field itself has to be provided by users. the function can give more dedicated message to tell users what to do.